direct_failure_detector/failure_detector.cc

```diff
@@ -478,7 +478,15 @@ static future<bool> ping_with_timeout(pinger::endpoint_id id, clock::timepoint_t
 
     auto f = pinger.ping(id, timeout_as);
     auto sleep_and_abort = [] (clock::timepoint_t timeout, abort_source& timeout_as, clock& c) -> future<> {
-        co_await c.sleep_until(timeout, timeout_as);
+        co_await c.sleep_until(timeout, timeout_as).then_wrapped([&timeout_as] (auto&& f) {
```

Comment:
Now that I look at this again I'm not sure if it'll work.
There is no guarantee that `sleep_until` won't throw before it yields (returns the future) - what if it throws that early? `then_wrapped` won't catch the exception.

For correctness we'd need to:
- specify at the `direct_failure_detector::clock` interface that `sleep_until` never throws directly, always returns an exceptional future instead of throwing
- verify that all implementations of `sleep_until` indeed satisfy this requirement

Reply:
It can't throw due timeout_as because timeout_as is aborted later.

If it throws due to timeout, it's a preexisting issue.

Reply:
Indeed